pkg/typesystem/typesystem.go

```diff
@@ -7,72 +7,54 @@ import (
 	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
 )
 
-type SchemaVersion int32
-
 const (
-	SchemaVersionUnspecified SchemaVersion = 0
-	SchemaVersion1_0         SchemaVersion = 1
-	SchemaVersion1_1         SchemaVersion = 2
+	SchemaVersion10 = "1.0"
+	SchemaVersion11 = "1.1"
 )
 
 var (
+	ErrDuplicateTypes       = errors.New("an authorization model cannot contain duplicate types")
 	ErrInvalidSchemaVersion = errors.New("invalid schema version")
 )
 
-func NewSchemaVersion(s string) (SchemaVersion, error) {
-	switch s {
-	case "", "1.0":
-		return SchemaVersion1_0, nil
-	case "1.1":
-		return SchemaVersion1_1, nil
-	default:
-		return SchemaVersionUnspecified, ErrInvalidSchemaVersion
-	}
-}
-
-func (v SchemaVersion) String() string {
-	switch v {
-	case SchemaVersion1_0:
-		return "1.0"
-	case SchemaVersion1_1:
-		return "1.1"
-	default:
-		return "unspecified"
-	}
-}
-
 type TypeSystem struct {
-	Version         SchemaVersion
+	version         string
 	TypeDefinitions map[string]*openfgapb.TypeDefinition
 }
 
-func NewTypeSystem(version SchemaVersion, typeDefinitions []*openfgapb.TypeDefinition) *TypeSystem {
+// New creates a *TypeSystem from an *openfgapb.AuthorizationModel. New assumes that the model
+// has already been validated.
+func New(model *openfgapb.AuthorizationModel) *TypeSystem {
 	tds := map[string]*openfgapb.TypeDefinition{}
-	for _, td := range typeDefinitions {
+	for _, td := range model.GetTypeDefinitions() {
 		tds[td.GetType()] = td
 	}
 
 	return &TypeSystem{
-		Version:         version,
+		version:         model.GetSchemaVersion(),
 		TypeDefinitions: tds,
 	}
 }
 
-// TODO: In a future PR refactor WriteAuthorizationModel to take a map[string]*openfgapb.TypeDefinition
-// and get rid of this method
-func (t *TypeSystem) GetTypeDefinitions() []*openfgapb.TypeDefinition {
-	tds := make([]*openfgapb.TypeDefinition, 0, len(t.TypeDefinitions))
-	for _, td := range t.TypeDefinitions {
-		tds = append(tds, td)
-	}
+func (t *TypeSystem) GetVersion() string {
```

Comment:
To not confuse this method with returning a version of the typesystem, could we name it schema_version? Schema version refers to the version of the model's schema whereas version is so generic I don't know if it refers to the type system's version or the model's version etc..

Reply:
Good point. Updated. Thanks!